compiler.md

```diff
@@ -1,155 +1,483 @@
+> faster future web.
 
-- is used for all code initially
-- generates code quickly without heavy optimizations
-- compilation with the base compiler is very fast generates little code
+- fast startup vs. peak performance
+- low memory vs. max optimization
+- Ignition Interpreter allows to run code with some amount of optimization very quickly and has
+  very low memory footprint
+- TurboFan makes functions that run a lot fast, sacrificing some memory in the process
+- designed to support entire JavaScript language and make it possible to quickly add new
+  features and to optimize them fast and incrementally
 
-### Runtime Profiler 
+## Simplified Pipeline
 
-- monitors the running system and identifies hot code
+[slide: pipeline 2010](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_163) |
+[slide: pipeline 2014](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_220) |
+[slide: pipeline 2016](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_249) |
+[slide: pipeline 2017](https://docs.google.com/presentation/d/1_eLlVzcj94_G4r9j9d_Lj5HRKFnq6jgpuPJtnmIBs88/edit#slide=id.g2134da681e_0_125)
 
-### Optimizing Compiler 
+Once crankshaft was taken out of the mix the below pipeline was possible
 
-- recompiles and optimizes hot code identified by the runtime profiler
-- uses static single assignment form to perform optimizations
-  - loop-invariant code motion
-  - linear-scan register allocation
-  - inlining.
-- optimization decisions are based on type information collected while running the code produced by the base compiler
+<img alt="simplified pipeline" src="http://benediktmeurer.de/images/2016/v8-new-pipeline-20161125.png" width="70%">
 
-### Deoptimization Support 
+### Basic Steps
 
-- allows the optimizing compiler to be optimistic in the assumptions it makes when generating code
-- deoptimization support allows to bail out to the code generated by the base compiler if the assumptions in the
-  optimized code turn out to be too optimistic
+1. Parse JavaScript into an [AST (abstract syntax tree)](https://en.wikipedia.org/wiki/Abstract_syntax_tree)
+2. Generate bytecode from that AST
+3. Turn bytecode into sequence of bytecodes by the BytecodeGenerator, which is part of the [Ignition Interpreter](https://v8project.blogspot.com/2016/08/firing-up-ignition-interpreter.html)
+  - sequences are divided on a per function basis
+4. Execute bytecode sequences via Ignition and collect feedback via inline caches
+  - feedback used by Ignition itself to speed up subsequent interpretation of the bytecode
+  - feedback used for speculative optimization by TurboFan when code is optimized
+5. _Speculatively_ optimize and compile bytecode using collected feedback to generate optimized machine code
+  for the current architecture
```

Comment:
Turbofan compiles and optimizes Ignition's bytecode to machine code for the current architecture, thereby using the collected feedback to make _speculative_ assumptions (i.e. TurboFan bakes in the assumption that a certain variable always holds a string).

Reply:
Again, this is just the summary.
Details on speculative optimization are found [here](https://github.com/thlorenz/v8-perf/blob/turbo-updates/compiler.md#speculative-optimization) including a code sample.